### PR TITLE
add granulate gmaestro addon

### PR DESCRIPTION
**Pull request by kirshenboim**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by kirshenboim** [comment]
@shapirov103 thank for your quick response, Ive fixed the implementation to match the guidelines you provided (published new [npm package](https://www.npmjs.com/package/@granulate/gmaestro-eks-blueprints-addon?activeTab=readme), added 3 secrets, opened a new pr in [cdk-eks-blueprints-patterns](https://github.com/aws-samples/cdk-eks-blueprints-patterns)).

File: docs/addons/gmaestro.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import {GmaestroAddOn} from '@granulate/gmaestro-eks-blueprints-addon';
+
+const app = new cdk.App();
+
+const addOn = new GmaestroAddOn({
+        clientIdSecretName: "<client id secret name>", // Create and copy from gMaestro deployment yaml
+        clientName: "<client name>",
+        clusterName: "<cluster name>",
+        namespace: "<namespace>",
+        grafanaMetricsSecretName: "<grafana metrics secret name>", // Create and copy from gMaestro deployment yaml
```
Comment (shapirov103) [inline docs/addons/gmaestro.md:42]: I am unclear what "Create and copy from gMaestro deployment yaml" means. Can you clarify in the doc before the snippet?

File: docs/addons/gmaestro.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+After a few seconds, you will gain full visibility into your K8s cluster objects.
+The first rightsizing recommendations may take up to 5 minutes to load.
+
+## `gMaestroAddOn` Required (props)
+Take the following parameter from the sample YAML file that was downloaded.
+
+#### `clientIdSecretName: string`
+
+Create a secret (as a plaintext) in AWS copy its value from the Deployment section `MAESTRO_CLIENT_ID`
```
Comment (shapirov103) [inline docs/addons/gmaestro.md:72]: Please specify in AWS Secrets Manager. There are other ways to inject the secret, i.e. through the parameter store.

**Review by shapirov103 — COMMENTED** [review]
@kirshenboim Looks great, getting there. A few mostly minor comments.

File: examples/blueprint-construct/index.ts [hunk 3]
```diff
@@ -146,15 +146,22 @@ export default class BlueprintConstruct {
                 notebookStack: 'jupyter/datascience-notebook',
                 values: { prePuller: { hook: { enabled: false }}}
             }),
-            new blueprints.EmrEksAddOn()
+            new blueprints.EmrEksAddOn(),
+            new blueprints.addons.GmaestroAddOn({
```
[the remark this answers is not in the document]
Comment (kirshenboim) [reply examples/blueprint-construct/index.ts]: done

File: lib/addons/gmaestro/index.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
+import {Construct} from 'constructs';
```
[the remark this answers is not in the document]
Comment (kirshenboim) [reply lib/addons/gmaestro/index.ts]: done

File: lib/addons/index.ts [hunk 5]
```diff
@@ -43,6 +43,7 @@ export * from './istio-base';
 export * from './istio-control-plane';
 export * from './jupyterhub';
 export * from './emr-on-eks';
+export * from './gmaestro';
```
[the remark this answers is not in the document]
Comment (kirshenboim) [reply lib/addons/index.ts]: done

File: docs/addons/gmaestro.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
+# gMaestro AddOn for Amazon EKS Blueprints
+
+This repository contains the source code for the gMaestro AddOn for [Amazon EKS Blueprints](https://aws-quickstart.github.io/cdk-eks-blueprints/). This AddOn is a [CDK](https://aws.amazon.com/cdk/) construct that allows customers to add gMaestro to their Amazon EKS clusters.
```
[the remark this answers is not in the document]
Comment (kirshenboim) [reply docs/addons/gmaestro.md]: done

File: docs/addons/gmaestro.md (hunk 1 above)
Comment (kirshenboim) [reply docs/addons/gmaestro.md:50]: done

File: docs/addons/gmaestro.md (hunk 2 above)
Comment (kirshenboim) [reply docs/addons/gmaestro.md]: done

**Review by shapirov103 — COMMENTED** [review]
@kirshenboim overall looks great. A couple of questions:

I noticed that you use `getSecretValue()` API to populate your secret, which is different (insecure) from the recommended approach outlined [here](https://aws-quickstart.github.io/cdk-eks-blueprints/extensibility/#passing-secrets-to-add-ons).

1. What was the example that you used to copy it? This method is only used for non-sensitive data. 
2. Could you review the provided section on secret handling and let me know how you would like proceed. 

The downside of using `getSecretValue` is due to the fact that these values are injected into the Cloudformation stack that backs the blueprint deployment. Such values will show up cleartext and will be accessible to anyone who has access to the CFN stack. Happy to discuss if needed. 

We could use the suggested approach, but the documentation will have to call out this vulnerability so that customers know about it.

**Comment by ShaharYak** [comment]
Hi @shapirov103, since @kirshenboim is OOO, I'll answer on his behalf.
Regarding the `grafanaLogsSecretName` and `grafanaMetricsSecretName`, currently, we pass those secret values hard coded. We are working to remove those values from the helm chart (we'll open a new pr in the aws-quickstart,cdk-eks-blueprints-patterns after the update).

Regarding the `client_id`, this value is already encrypted from our side. 
Also, attaching the PR for aws-samples repository.
https://github.com/aws-samples/cdk-eks-blueprints-patterns/pull/57 (edited)

**Comment by shapirov103** [comment]
@ShaharYak 

If you would like to move forward with the current implementation then docs should be updated to reflect that the approach has a security vulnerability so that customers are aware. 

You mentioned client_id is already encrypted, can't this value be reused as is elsewhere? E.g. a malicious actor gains access to the encrypted client secret and uses it in a different blueprint for example?

Happy to discuss if needed.

**Comment by kirshenboim** [comment]
Hi @shapirov this is understandable. We added an explanation in the docs as per your request. We will open a new PR in the following weeks that will include your suggested implementation and remove the explanation from the docs.

File: docs/addons/gmaestro.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+The secret name from the Prerequisite section 3.ii.
+
+#### `grafanaLogsSecretName: string`
+
+The secret name from the Prerequisite section 3.iii.
+
+
+## Security issue
+
+As the implementation uses getSecretValue(), the secret parameters will be displayed as plain text in the CloudFormation. These secrets are gMaestro’s secrets which don’t affect the user.
```
Comment (shapirov103) [inline docs/addons/gmaestro.md:105]: Many customers have no idea what `getSecretValue()` means in this context.

Let's state the following:

1. The implementation requires access to the AWS Secrets Manager at build time to retrieve secret values.
2. The secret values will be stored as plain text in the resulting CloudFormation stack, meaning that any user with access to view CloudFormation stack can gain access to these secrets.

Note: These secrets are specific to gMaestro and don't affect customer account beyond the scope of the gMaestro add-on.

My wording can surely be improved further, I was just attempting to convey the point, please feel free to adjust.

**Review by shapirov103 — COMMENTED** [review]
@kirshenboim thank you for your changes. The PR looks good and about ready to merge. I added one comment, and I will validate the pattern as the next step. 
Most of us were out on conference this week (internal AWS), hence the delayed response.

**Comment by kirshenboim** [comment]
@shapirov103 updated Security issue text

**Comment by shapirov103** [comment]
@kirshenboim I will release 1.6 version this week and test out the pattern on it. Will merge, provided the pattern succeeds.

**Comment by elamaran11** [comment]
@kirshenboim @noamsalinger We dont see any traction on this addon and i also see `conflicts` which needs to be resolved. Could you please check on this, fix all that is required so we can move forward with this addon.?

**Comment by marina-vrublevsky** [comment]
Hi @elamaran11 @shapirov103 I've fixed the conflicts of this PR in https://github.com/aws-quickstart/cdk-eks-blueprints/pull/727.

**Comment by elamaran11** [comment]
Hi @marina-vrublevsky We have feedback on the Blueprints Pattern repo too, please fix those. We can dive in to this PR, once you fix the pattern and once we cover validation. 

https://github.com/aws-samples/cdk-eks-blueprints-patterns/pull/57

**Comment by elamaran11** [comment]
@marina-vrublevsky Closing this PR as we have a new PR for Granulate.

**Pull request closed without merging**